eo-maven-plugin/src/main/java/org/eolang/maven/Timeout.java

```diff
@@ -0,0 +1,139 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.eolang.maven;
+
+import com.jcabi.log.Logger;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Timeout for execution thread.
+ *
+ * @since 0.28.12
+ */
+final class Timeout {
+
+    /**
+     * Timeout value that determines how long to wait before interrupt.
+     */
+    private final int value;
+
+    /**
+     * Timeout units.
+     */
+    private final TimeUnit unit;
+
+    /**
+     * Thread to interrupt if timeout limit reached.
+     */
+    private final Thread thread;
+
+    /**
+     * Inner flag that purpose to determine if timeout have to interrupt a thread.
+     */
+    private final AtomicBoolean finish;
+
+    /**
+     * The shortest constructor with default preset for units and thread.
+     *
+     * @param value How long to wait before interrupt.
+     */
+    Timeout(final int value) {
+        this(value, Thread.currentThread());
+    }
+
+    /**
+     * Simple constructor with timeout and time units.
+     * Useful for tests.
+     *
+     * @param value How long to wait before interrupt.
+     * @param unit Time units
+     */
+    Timeout(final int value, final TimeUnit unit) {
+        this(value, unit, Thread.currentThread());
+    }
+
+    /**
+     * Simple constructor with timeout and thread.
+     *
+     * @param value How long to wait before interrupt.
+     * @param thread To interrupt if timeout is reached.
+     */
+    private Timeout(final int value, final Thread thread) {
+        this(value, TimeUnit.SECONDS, thread);
+    }
+
+    /**
+     * The main constructor.
+     *
+     * @param value How long to wait before interrupt.
+     * @param unit Time units
+     * @param thread To interrupt if timeout is reached.
+     */
+    private Timeout(final int value, final TimeUnit unit, final Thread thread) {
+        this.value = value;
+        this.unit = unit;
+        this.thread = thread;
+        this.finish = new AtomicBoolean(false);
+    }
+
+    /**
+     * Starts counting down before interrupting a thread.
+     * If timeout limit is reached the thread will be interrupted.
+     */
+    void start() {
+        final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+        executor.schedule(this::interrupt, this.value, this.unit);
+        executor.shutdown();
+    }
+
+    /**
+     * Stops counting down before interrupting a thread.
+     * If that method was call before timeout limit is reached then nothing will happen and
+     * normal execution will continue.
+     */
+    void stop() {
+        this.finish.set(true);
+    }
+
+    /**
+     * Interrupts thread if timeout wasn't sopped.
+     */
+    private void interrupt() {
+        if (!this.finish.get()) {
```

Comment:
@volodya-lombrozo this block is not synchronized, this is the problem. You check for the flag and then interrupt the thread. But the flag may be reset to TRUE right after you check it. Then, you will interrupt the thread that is already stopped.

Reply:
> @volodya-lombrozo how about this:
> 
> ```java
> this.waitAndInterrupt(Thread.currentThread(), 5000);
> 
> void waitAndInterrupt(Thread t, int msec) {
>   new Thread() {
>     @Override
>     public void run() {
>       Thread.sleep(msec);
>       synchronized (t) {
>         if (t.isAlive()) {
>           t.interrupt();
>         }
>       }
>     }
>   }.start();
> }
> ```
> 
> Keep in mind, the shorter the code, the better. Short code is easier to maintain. Every extra class is a future pain.

@yegor256 Done. Please, check